lib/rules/declarations-for-function-components.js

```diff
@@ -0,0 +1,118 @@
+/**
+ * @fileoverview Function declarations for functional components
+ * @author Stefan Wullems
+ */
+
+'use strict';
+
+const Components = require('../util/Components');
+const docsUrl = require('../util/docsUrl');
+
+// ------------------------------------------------------------------------------
+// Rule Definition
+// ------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Function declarations for functional components',
+      category: 'Best Practices',
+      recommended: false,
+      url: docsUrl('declarations-for-function-components')
+    },
+
+    schema: []
+  },
+
+  create: Components.detect((context, components) => {
+    const NOT_DECLARED_FUNCTION_COMPONENT_MESSAGE = 'Function component is not a function declaration';
+
+    /**
+     * Mark a component as being a declared function component
+     * @param {ASTNode} node The AST node being checked.
+     */
+    function markComponentAsDeclaredFunctionComponent(node) {
+      components.set(node, {
+        isDeclaredFunctionalComponent: true
+      });
+    }
+
+    /**
+     * Mark a component as being a declared function component
+     * @param {ASTNode} node The AST node being checked.
+     */
+    function markAsMethodDefinition(node) {
+      components.set(node, {
+        isMethodDefinition: true
+      });
+    }
+
+    /**
+     * Mark a component as being a function component
+     * @param {ASTNode} node The AST node being checked.
+     */
+    function markComponentAsFunctionalComponent(node) {
+      components.set(node, {
+        isFunctionalComponent: true
+      });
+    }
+
+    /**
+     * Reports that component is defined with an arrow function
+     * @param {Object} component The component to process
+     */
+    function reportNotDeclaredFunctionalComponent(component) {
+      context.report({
+        node: component.node,
+        message: NOT_DECLARED_FUNCTION_COMPONENT_MESSAGE,
+        data: {
+          component: component.name
+        }
+      });
+    }
+
+    // --------------------------------------------------------------------------
+    // Public
+    // --------------------------------------------------------------------------
+
+    return {
+
+      MethodDefinition(node) {
+        if (components.get(node)) {
+          markAsMethodDefinition(node);
+        }
+      },
+
+      FunctionExpression(node) {
+        if (components.get(node)) {
+          markComponentAsFunctionalComponent(node);
+        }
+      },
+
+      ArrowFunctionExpression(node) {
+        if (components.get(node)) {
+          markComponentAsFunctionalComponent(node);
+        }
+      },
+
+      FunctionDeclaration(node) {
+        if (components.get(node)) {
+          markComponentAsFunctionalComponent(node);
+          markComponentAsDeclaredFunctionComponent(node);
```

Comment:
why does `markComponentAsDeclaredFunctionComponent` not implicitly call `markComponentAsFunctionalComponent`?